DL-124: Use Java8 Future rather than twitter Future

sijie opened the pull request:
Switch to use Java8 CompletableFuture, to reduce dependencies introduced by twitter future and make it more friendly to users (users don't think of using which version of scala).

This change is based on #132 . Gitsha ce0686e is the change to review.

The changes:

- Change Future to CompletableFuture
- Map to thenApply
- flatMap to thenCompose
- Added a FutureEventListener, and switch addEvenListener to whenComplete (or whenCompleteAsync)
- setValue to complete
- setException to completeExceptionally
- add rescue, ignore, ensure to FutureUtils as util functions.

sijie commented:
tag @fcuny @mgodave @leighst @jiazhai for reviews

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKAbstractLogWriter.java:
```diff
@@ -357,80 +354,72 @@ private void truncateLogSegmentsIfNecessary(BKLogWriteHandler writeHandler) {
 
-    private Future<BKLogSegmentWriter> asyncStartNewLogSegment(final BKLogWriteHandler writeHandler,
+    private CompletableFuture<BKLogSegmentWriter> asyncStartNewLogSegment(final BKLogWriteHandler writeHandler,
                                                                final long startTxId,
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKAsyncLogReader.java:
```diff
@@ -381,16 +369,16 @@ public String getStreamName() {
     @Override
-    public synchronized Future<List<LogRecordWithDLSN>> readBulk(int numEntries,
+    public synchronized CompletableFuture<List<LogRecordWithDLSN>> readBulk(int numEntries,
                                                                  long waitTime,
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKAsyncLogReader.java:
```diff
@@ -404,7 +392,7 @@ public String getStreamName() {
      */
-    private synchronized Future<List<LogRecordWithDLSN>> readInternal(int numEntries,
+    private synchronized CompletableFuture<List<LogRecordWithDLSN>> readInternal(int numEntries,
                                                                       long deadlineTime,
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKAsyncLogWriter.java:
```diff
@@ -206,7 +197,7 @@ private BKLogSegmentWriter getCachedLogSegmentWriter() throws WriteException {
 
-    private Future<BKLogSegmentWriter> getLogSegmentWriter(long firstTxid,
+    private CompletableFuture<BKLogSegmentWriter> getLogSegmentWriter(long firstTxid,
                                                            boolean bestEffort,
```
Please also do the code alignment change in this file, at line 201, 212, 242, 272,

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKDistributedLogManager.java:
```diff
@@ -525,75 +495,63 @@ public BKSyncLogWriter startLogSegmentNonPartitioned() throws IOException {
 
-    private Future<DLSN> getDLSNNotLessThanTxId(long fromTxnId,
+    private CompletableFuture<DLSN> getDLSNNotLessThanTxId(long fromTxnId,
                                                 final List<LogSegmentMetadata> segments) {
```
Please also do the code alignment change in this file, also at line 565, 707, 838.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKLogHandler.java:
```diff
@@ -397,15 +398,15 @@ private Long sum(List<Long> values) {
 
-    public Future<LogRecordWithDLSN> asyncReadLastRecord(final LogSegmentMetadata l,
+    public CompletableFuture<LogRecordWithDLSN> asyncReadLastRecord(final LogSegmentMetadata l,
                                                          final boolean fence,
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKLogWriteHandler.java:
```diff
@@ -483,23 +482,23 @@ protected long assignLogSegmentSequenceNumber() throws IOException {
 
-    protected Future<BKLogSegmentWriter> asyncStartLogSegment(final long txId,
+    protected CompletableFuture<BKLogSegmentWriter> asyncStartLogSegment(final long txId,
                                                               final boolean bestEffort,
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/BookKeeperClient.java:
```diff
@@ -198,52 +196,52 @@ public synchronized BookKeeper get() throws IOException {
 
-    public Future<Void> deleteLedger(long lid,
+    public CompletableFuture<Void> deleteLedger(long lid,
                                      final boolean ignoreNonExistentLedger) {
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/bk/SimpleLedgerAllocator.java:
```diff
@@ -105,73 +103,71 @@ public ConcurrentObtainException(Phase phase, String msg) {
     private final QuorumConfigProvider quorumConfigProvider;
 
-    static Future<Versioned<byte[]>> getAndCreateAllocationData(final String allocatePath,
+    static CompletableFuture<Versioned<byte[]>> getAndCreateAllocationData(final String allocatePath,
```
Please also do the code alignment change here, and also at line 128, 159.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/impl/ZKLogSegmentMetadataStore.java:
```diff
@@ -350,40 +349,40 @@ public void process(WatchedEvent event) {
     @Override
-    public Future<Versioned<List<String>>> getLogSegmentNames(String logSegmentsPath,
+    public CompletableFuture<Versioned<List<String>>> getLogSegmentNames(String logSegmentsPath,
                                                               LogSegmentNamesListener listener) {
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/impl/logsegment/BKLogSegmentEntryStore.java:
```diff
@@ -186,13 +185,13 @@ LedgerAllocator createLedgerAllocator(LogMetadataForWriter logMetadata,
     @Override
-    public Future<LogSegmentEntryReader> openReader(LogSegmentMetadata segment,
+    public CompletableFuture<LogSegmentEntryReader> openReader(LogSegmentMetadata segment,
                                                     long startEntryId) {
```
Please also do the code alignment change here, also at line 244.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/impl/metadata/ZKLogStreamMetadataStore.java:
```diff
@@ -237,15 +227,13 @@ public DistributedLock createWriteLock(LogMetadataForWriter metadata) {
 
-    private Future<Void> ensureReadLockPathExist(final LogMetadata logMetadata,
+    private CompletableFuture<Void> ensureReadLockPathExist(final LogMetadata logMetadata,
                                                  final String readLockPath) {
```
Please also do the code alignment change here, and line 269, 312, 521, 572.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/lock/ZKSessionLock.java:
```diff
@@ -1088,9 +1063,9 @@ public String getActionName() {
 
-    private Future<String> checkLockOwnerAndWaitIfPossible(final LockWatcher lockWatcher,
+    private CompletableFuture<String> checkLockOwnerAndWaitIfPossible(final LockWatcher lockWatcher,
                                                            final boolean wait) {
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/logsegment/LogSegmentEntryStore.java:
```diff
@@ -58,7 +57,7 @@
      */
-    Future<LogSegmentEntryReader> openReader(LogSegmentMetadata segment,
+    CompletableFuture<LogSegmentEntryReader> openReader(LogSegmentMetadata segment,
                                              long startEntryId);
```
Please also do the code alignment change here, and line 71

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/metadata/LogSegmentMetadataStoreUpdater.java:
```diff
@@ -73,7 +72,7 @@ private String formatLogSegmentSequenceNumber(long logSegmentSeqNo) {
     @Override
-    public Future<LogSegmentMetadata> changeSequenceNumber(LogSegmentMetadata segment,
+    public CompletableFuture<LogSegmentMetadata> changeSequenceNumber(LogSegmentMetadata segment,
                                                            long logSegmentSeqNo) {
```
Please also do the code alignment change here.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/metadata/MetadataUpdater.java:
```diff
@@ -56,7 +56,7 @@
      */
-    Future<LogSegmentMetadata> changeSequenceNumber(LogSegmentMetadata segment,
+    CompletableFuture<LogSegmentMetadata> changeSequenceNumber(LogSegmentMetadata segment,
                                                     long logSegmentSeqNo);
```
Please also do the code alignment change here, and line 102.

jiazhai commented on distributedlog-core/src/main/java/org/apache/distributedlog/util/Utils.java:
```diff
@@ -445,35 +440,35 @@ public void processResult(int rc, String path, Object ctx) {
 
-    public static Future<Void> asyncClose(@Nullable AsyncCloseable closeable,
+    public static CompletableFuture<Void> asyncClose(@Nullable AsyncCloseable closeable,
                                           boolean swallowIOException) {
```
Please also do the code alignment change here, and line 573, 586.

leighst commented on distributedlog-benchmark/pom.xml:
```diff
@@ -27,12 +27,12 @@
     <dependency>
       <groupId>org.apache.distributedlog</groupId>
-      <artifactId>distributedlog-client</artifactId>
+      <artifactId>distributedlog-proxy-client</artifactId>
```
why?

sijie replied:
this change was based on #132. github doesn't allow pull request based on the other branch.

leighst commented on distributedlog-core/src/main/java/org/apache/distributedlog/BKLogReadHandler.java:
```diff
@@ -201,33 +183,31 @@ public void safeRun() {
-                FutureUtils.cancel(acquireFuture);
-                return null;
+        final CompletableFuture<Void> threadAcquirePromise = new CompletableFuture<Void>();
+        threadAcquirePromise.whenComplete((value, cause) -> {
```
do i understand correctly java8 future doesn't really have the concept of interrupt?

sijie replied:
correctly, java8 future doesn't have the concept of interrupt. I try to simulate the similar thing.

leighst commented on distributedlog-core/src/main/java/org/apache/distributedlog/util/Utils.java:
```diff
@@ -635,4 +631,80 @@ public static String getParent(final String path) {
+            + BKException.getMessage(transmitResult), transmitResult);
+    }
+
+    public static <T> T ioResult(CompletableFuture<T> result) throws IOException {
```
why not have a java8 version of FutureUtils? 
seems weird to throw all of this into one file

leighst commented:
Is there some way we can make migration easier? We're going to have to do a lot of translation to twitter futures when we integrate. Should we be thinking about some kind of adapter layer?

sijie commented:
@leighst : let's discuss this tomorrow face to face. I am thinking of a new api with java8 future that works for both core-library and proxy. that means the way to distinguish core and library is just the uri. we can have the old twitter-future based api wrapper over the new java8 future api. so it will be backward compatible.

sijie commented:
@leighst - I pushed a new change:

- moved the interfaces to org.apache.distributedlog.api, which they are using Java8 interface.
- moved the old interfaces to distributedlog-core-twitter module, they are are backward compatible to old application.

So the old application can use distributedlog-core-twitter and keep the binary compatible.

sijie commented:
Let me know if this approach works for you.

sijie commented on distributedlog-benchmark/conf/log4j.properties:
```diff
@@ -30,11 +30,7 @@ log4j.logger.org.apache.zookeeper=INFO
-log4j.logger.org.apache.distributedlog.util.MonitoredFuturePool=INFO, Executors
-log4j.logger.org.apache.distributedlog.util.MonitoredScheduledThreadPoolExecutor=INFO, Executors
 log4j.logger.org.apache.bookkeeper.util.SafeRunnable=INFO, Executors
-log4j.additivity.org.apache.distributedlog.util.MonitoredFuturePool=false
```
MonitoredFuturePool and MonitoredScheduledThreadPoolExecutor are already removed in this patch. because in Java8, you can configure a thread where the callbacks will be executed, there is no need for additional future pool, we just use the ordered scheduler.
